### ISSUE TITLE
Qt5 blocker: DB4S + Qt5 + OSX == not all data shown

### ISSUE THREAD

**Issue by justinclift** [sqlitebrowser/sqlitebrowser#750]
### Details for the issue

When built with Qt5 on OSX, DB4S isn't showing all of the table data.

The table info shows the correct total number of records, but the scroll bar isn't correctly scrolling to the bottom of them, and the page buttons don't work properly either.

It's a bit hard to describe using just words, so I'll do some screenrecording later today to show the problem in action.

This doesn't happen on Windows, and it also doesn't happen for Qt4 on OSX.  So, there's possibly something OSX specific about the problem.

This is a blocker to going Qt5 only. 😦 

#### I'm opening this issue because:

- [ ] DB4S is crashing
- [x] DB4S has a bug
- [ ] DB4S needs a feature
- [ ] DB4S has another problem

#### I'm using DB4S on:

- [ ] Windows
- [ ] Linux: ( _distro:_ ___ )
- [x] Mac OS
- [ ] Other: ___

**Comment by MarqueIV** [comment]
Hi Justin! This was the same issue I described to you earlier.  As mentioned, while in QT4 you can see all the data, the paging controls are wonky there too (i.e. I don't know of *any* version where they work correctly.)

**Comment by justinclift** [comment]
Yep, this is the one.  Figured it best to put in a separate issue, so we can track this one down as it's own thing.

Which reminds me, this really needs some screenrecording things of the bug in action to properly convey the problem.

Don't suppose you'd have the time/interest to do some?  If not, no worries, I will get around to it. :smile:

**Comment by MarqueIV** [comment]
[OddPagingInDbBrowser.mov.zip](https://github.com/sqlitebrowser/sqlitebrowser/files/446295/OddPagingInDbBrowser.mov.zip)
Movie attached.  You can see that while only eight rows are shown (sizing changes this amount), it's still paging by up to 100 places, meaning if you want to see anything between 9 and 90, you can't use the paging controls.  Paging size should match the shown number of rows, always.

**Comment by justinclift** [comment]
Ahhh.  I'm suspecting this is a different issue, though probably related.  But, both should be looked into.

**Comment by MarqueIV** [comment]
Yes, different than the QT5 issue of not seeing all the rows, which I did see as well. This issue exists in both QT4 and QT5 versions.

**Comment by MKleusberg** [comment]
@justinclift Did you have a chance to do a screen recording of the scrolling behaviour yet?
Also, does this happen for each table in every database, or just for some of them? Does it depend on the number of rows in a table or the size of the window (i.e. the number of rows shown on the screen)? Is it sort of random or are always the same rows missing?

Lots of questions, I know :wink: But I'm trying to narrow this down for easiest debugging since I don't have a OS X system available here.

**Comment by justinclift** [comment]
Ahhh yeah, I really need to get around to that.

It happens for every table in the database, regardless of # of rows, table size, window size, etc.

My suspicion is it's something to do with the way we're paging things in the backend.  But no more idea than that.

Give me a bit, I'll compile it again and record it now.

**Comment by justinclift** [comment]
k, screen recording here:

&nbsp; &nbsp; http://nightlies.sqlitebrowser.org/issue_uploads/DB4S_Issue_750v1.mp4

**Comment by justinclift** [comment]
There is only one table in the database, so I wasn't able to show it with multiple tables, but it should be clear enough to show the problem.

Can provide the database too if that helps. :smile:

**Comment by MKleusberg** [comment]
Ah interesting :smiley: Not entirely sure how to track this down though. Can you do me a favour and tell me the value in Preferences -> Database -> Prefetch block size as well as recording this again with the SQL log opened?

**Comment by justinclift** [comment]
Yep, can do.  1 min. :smile:

**Comment by justinclift** [comment]
@MKleusberg As a thought towards debugging, maybe we should add some println() style statements to the SQL Log, showing the values of whatever seems appropriate?  I'm completely ok with compiling stuff and recording again.  I just don't know which things we'd be interested in. :wink:

**Comment by MKleusberg** [comment]
@justinclift Yes, let's do that :smiley: To be honest, I still don't see the source of the problem, though it definitely feels like we're getting closer.

Can you try this hacky patch? Just scroll all the way down and paste the SQL log. No screen record needed this time :smiley: 
```diff
diff --git a/src/ExtendedTableWidget.cpp b/src/ExtendedTableWidget.cpp
index 42c9b1f..aa202b9 100644
--- a/src/ExtendedTableWidget.cpp
+++ b/src/ExtendedTableWidget.cpp
@@ -330,6 +330,8 @@ void ExtendedTableWidget::updateGeometries()
     }
 }
 
+#include "MainWindow.h"
+
 void ExtendedTableWidget::vscrollbarChanged(int value)
 {
     // Cancel if there is no model set yet - this shouldn't happen (because without a model there should be no scrollbar) but just to be sure...
@@ -337,8 +339,12 @@ void ExtendedTableWidget::vscrollbarChanged(int value)
         return;
 
     // Fetch more data from the DB if necessary
-    if((value + numVisibleRows()) >= model()->rowCount() && model()->canFetchMore(QModelIndex()))
+    int nvr = numVisibleRows();
+    if((value + nvr) >= model()->rowCount() && model()->canFetchMore(QModelIndex()))
+    {
+        dynamic_cast<MainWindow*>(model()->parent())->logSql(QString("ExtendedTableWidget::vscrollbarChanged -> value=%1, numVisibleRows=%2, modelRowCount=%3.").arg(value).arg(nvr).arg(model()->rowCount()), kLogMsg_App);
         model()->fetchMore(QModelIndex());
+    }
 }
 
 int ExtendedTableWidget::numVisibleRows()
```

**Comment by justinclift** [comment]
Done. :smile:

```
PRAGMA foreign_keys = "1";
SELECT type,name,sql,tbl_name FROM sqlite_master UNION SELECT type,name,sql,tbl_name FROM sqlite_temp_master;
PRAGMA encoding
SELECT COUNT(*) FROM (SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC);
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 0, 50000;
ExtendedTableWidget::vscrollbarChanged -> value=50093, numVisibleRows=15, modelRowCount=50000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 50000, 50000;
ExtendedTableWidget::vscrollbarChanged -> value=100186, numVisibleRows=14, modelRowCount=100000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 100000, 50000;
ExtendedTableWidget::vscrollbarChanged -> value=151501, numVisibleRows=14, modelRowCount=150000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 150000, 50000;
ExtendedTableWidget::vscrollbarChanged -> value=201594, numVisibleRows=15, modelRowCount=200000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 200000, 50000;
ExtendedTableWidget::vscrollbarChanged -> value=259018, numVisibleRows=15, modelRowCount=250000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 250000, 50000;
ExtendedTableWidget::vscrollbarChanged -> value=300558, numVisibleRows=14, modelRowCount=300000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 300000, 50000;
ExtendedTableWidget::vscrollbarChanged -> value=351874, numVisibleRows=14, modelRowCount=350000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 350000, 50000;
ExtendedTableWidget::vscrollbarChanged -> value=401967, numVisibleRows=15, modelRowCount=400000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 400000, 50000;
ExtendedTableWidget::vscrollbarChanged -> value=454503, numVisibleRows=14, modelRowCount=450000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 450000, 50000;
```

**Comment by justinclift** [comment]
It seems like the table inside the enclosing widget knows the full table size.  It's just the outside scroll thing seems clueless. 😦

**Comment by MKleusberg** [comment]
Yeah, the problem seems to be with the scrollbar itself :confused: Let's do some further testing. Here's an updated patch. It need's to be applied on a clean head, i.e. not on the last patch. Again, I only need the SQL log.

```diff
diff --git a/src/ExtendedTableWidget.cpp b/src/ExtendedTableWidget.cpp
index 42c9b1f..3504ec4 100644
--- a/src/ExtendedTableWidget.cpp
+++ b/src/ExtendedTableWidget.cpp
@@ -315,6 +315,8 @@ void ExtendedTableWidget::keyPressEvent(QKeyEvent* event)
         QTableView::keyPressEvent(event);
 }
 
+#include "MainWindow.h"
+
 void ExtendedTableWidget::updateGeometries()
 {
     // Call the parent implementation first - it does most of the actual logic
@@ -326,7 +328,10 @@ void ExtendedTableWidget::updateGeometries()
         // If so and if it is a SqliteTableModel and if the parent implementation of this method decided that a scrollbar is needed, update its maximum value
         SqliteTableModel* m = qobject_cast<SqliteTableModel*>(model());
         if(m && verticalScrollBar()->maximum())
+        {
+            dynamic_cast<MainWindow*>(model()->parent())->logSql(QString("ExtendedTableWidget::updateGeometries -> totalRowCount=%1, numVisibleRows=%2.").arg(m->totalRowCount()).arg(numVisibleRows()), kLogMsg_App);
             verticalScrollBar()->setMaximum(m->totalRowCount() - numVisibleRows() + 1);
+        }
     }
 }
 
@@ -337,8 +342,12 @@ void ExtendedTableWidget::vscrollbarChanged(int value)
         return;
 
     // Fetch more data from the DB if necessary
-    if((value + numVisibleRows()) >= model()->rowCount() && model()->canFetchMore(QModelIndex()))
+    int nvr = numVisibleRows();
+    if((value + nvr) >= model()->rowCount() && model()->canFetchMore(QModelIndex()))
+    {
+        dynamic_cast<MainWindow*>(model()->parent())->logSql(QString("ExtendedTableWidget::vscrollbarChanged -> value=%1, numVisibleRows=%2, modelRowCount=%3.").arg(value).arg(nvr).arg(model()->rowCount()), kLogMsg_App);
         model()->fetchMore(QModelIndex());
+    }
 }
 
 int ExtendedTableWidget::numVisibleRows()
```

**Comment by justinclift** [comment]
Sure. :smile:

```
PRAGMA foreign_keys = "1";
SELECT type,name,sql,tbl_name FROM sqlite_master UNION SELECT type,name,sql,tbl_name FROM sqlite_temp_master;
PRAGMA encoding
SELECT COUNT(*) FROM (SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC);
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 0, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=51315, numVisibleRows=14, modelRowCount=50000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 50000, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=101408, numVisibleRows=14, modelRowCount=100000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 100000, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=150280, numVisibleRows=14, modelRowCount=150000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 150000, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=200373, numVisibleRows=14, modelRowCount=200000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 200000, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=250466, numVisibleRows=15, modelRowCount=250000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 250000, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=15.
ExtendedTableWidget::vscrollbarChanged -> value=301780, numVisibleRows=14, modelRowCount=300000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 300000, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=351874, numVisibleRows=14, modelRowCount=350000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 350000, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=403189, numVisibleRows=14, modelRowCount=400000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 400000, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=450839, numVisibleRows=15, modelRowCount=450000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 450000, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=15.
```

**Comment by justinclift** [comment]
Hmmm, those row values in the bottom left.  eg: `16657 - 16670 of 499723`

Looking at the .ui and source, that "to" value (the middle one), seems to be getting involved with the false maximum:

* https://github.com/sqlitebrowser/sqlitebrowser/blob/969e263e5bf71e63df8ee908cb4a2014b9a55c9b/src/MainWindow.cpp#L680

The code for that line is:

```cpp
int to = ui->dataTable->verticalHeader()->visualIndexAt(ui->dataTable->height()) - 1;
```

Perhaps either `dataTable->height()` is returning the wrong value, or the visualIndexAt() method isn't getting things correct.  Maybe some println() from that would help? :smile:

**Comment by MKleusberg** [comment]
I think the code in MainWindow.cpp is working correctly. All it does is getting the row numbers of the first and the last visible row in the table view. And as far as I can tell from your videos it's doing that correctly. So the problem must be in the table view somewhere and the fale maximum value at the bottom is just an effect of that.

The scrollbar is also being set as expected. Looking at your screen recordings now, it seems like the scrollbar is doing exactly what it's supposed to do, triggering prefetches where it's at least plausible to do so. But the table view seems to be lagging and getting out of sync.

**Comment by justinclift** [comment]
You're probably right.  If it helps though, this is the output from scrolling to the end of the table, using this [simple patch](https://github.com/justinclift/sqlitebrowser/commit/def049a8d314f8e4f6cd5fdfca76c6d6077ffcfc) on top of your 2nd one:

```
dataTableHeight = 30, to = -2
PRAGMA foreign_keys = "1";
SELECT type,name,sql,tbl_name FROM sqlite_master UNION SELECT type,name,sql,tbl_name FROM sqlite_temp_master;
PRAGMA encoding
SELECT COUNT(*) FROM (SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC);
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 0, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
dataTableHeight = 429, to = 13
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
dataTableHeight = 429, to = 94
dataTableHeight = 429, to = 420
dataTableHeight = 429, to = 502
dataTableHeight = 429, to = 583
dataTableHeight = 429, to = 664
dataTableHeight = 429, to = 746
dataTableHeight = 429, to = 827
dataTableHeight = 429, to = 868
dataTableHeight = 429, to = 950
dataTableHeight = 429, to = 1031
dataTableHeight = 429, to = 1112
dataTableHeight = 429, to = 1194
dataTableHeight = 429, to = 1275
dataTableHeight = 429, to = 1357
dataTableHeight = 429, to = 1479
dataTableHeight = 429, to = 1560
ExtendedTableWidget::vscrollbarChanged -> value=51315, numVisibleRows=14, modelRowCount=50000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 50000, 50000;
dataTableHeight = 429, to = 1723
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
dataTableHeight = 429, to = 2049
dataTableHeight = 429, to = 2131
dataTableHeight = 429, to = 2171
dataTableHeight = 429, to = 2253
dataTableHeight = 429, to = 2334
dataTableHeight = 429, to = 2416
dataTableHeight = 429, to = 2497
dataTableHeight = 429, to = 2579
dataTableHeight = 429, to = 2660
dataTableHeight = 429, to = 2782
dataTableHeight = 429, to = 2864
dataTableHeight = 429, to = 3027
dataTableHeight = 429, to = 3108
dataTableHeight = 429, to = 3189
dataTableHeight = 429, to = 3271
ExtendedTableWidget::vscrollbarChanged -> value=100186, numVisibleRows=14, modelRowCount=100000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 100000, 50000;
dataTableHeight = 429, to = 3352
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
dataTableHeight = 429, to = 3719
dataTableHeight = 429, to = 3882
dataTableHeight = 429, to = 3963
dataTableHeight = 429, to = 4004
dataTableHeight = 429, to = 4085
dataTableHeight = 429, to = 4248
dataTableHeight = 429, to = 4330
dataTableHeight = 429, to = 4493
dataTableHeight = 429, to = 4574
dataTableHeight = 429, to = 4696
dataTableHeight = 429, to = 4859
dataTableHeight = 429, to = 4941
ExtendedTableWidget::vscrollbarChanged -> value=150280, numVisibleRows=14, modelRowCount=150000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 150000, 50000;
dataTableHeight = 429, to = 5022
dataTableHeight = 429, to = 5104
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=15.
dataTableHeight = 429, to = 6081
dataTableHeight = 429, to = 6325
dataTableHeight = 429, to = 6407
dataTableHeight = 429, to = 6610
ExtendedTableWidget::vscrollbarChanged -> value=202816, numVisibleRows=14, modelRowCount=200000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 200000, 50000;
dataTableHeight = 429, to = 6773
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
dataTableHeight = 429, to = 7058
dataTableHeight = 429, to = 7832
dataTableHeight = 429, to = 7954
dataTableHeight = 429, to = 8199
ExtendedTableWidget::vscrollbarChanged -> value=252909, numVisibleRows=14, modelRowCount=250000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 250000, 50000;
dataTableHeight = 429, to = 8443
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
dataTableHeight = 429, to = 8647
dataTableHeight = 429, to = 8728
dataTableHeight = 429, to = 8891
dataTableHeight = 429, to = 9135
dataTableHeight = 429, to = 9258
dataTableHeight = 429, to = 9543
dataTableHeight = 429, to = 9787
dataTableHeight = 429, to = 9828
dataTableHeight = 429, to = 9909
dataTableHeight = 429, to = 9991
ExtendedTableWidget::vscrollbarChanged -> value=301781, numVisibleRows=14, modelRowCount=300000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 300000, 50000;
dataTableHeight = 429, to = 10072
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
dataTableHeight = 429, to = 10520
dataTableHeight = 429, to = 10765
dataTableHeight = 429, to = 10846
dataTableHeight = 429, to = 11009
dataTableHeight = 429, to = 11090
dataTableHeight = 429, to = 11213
dataTableHeight = 429, to = 11294
dataTableHeight = 429, to = 11375
dataTableHeight = 429, to = 11579
ExtendedTableWidget::vscrollbarChanged -> value=350652, numVisibleRows=14, modelRowCount=350000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 350000, 50000;
dataTableHeight = 429, to = 11701
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
dataTableHeight = 429, to = 12108
dataTableHeight = 429, to = 12271
dataTableHeight = 429, to = 12312
dataTableHeight = 429, to = 12394
dataTableHeight = 429, to = 12556
dataTableHeight = 429, to = 12638
dataTableHeight = 429, to = 12719
dataTableHeight = 429, to = 12801
dataTableHeight = 429, to = 12964
dataTableHeight = 429, to = 13004
dataTableHeight = 429, to = 13086
dataTableHeight = 429, to = 13167
dataTableHeight = 429, to = 13330
ExtendedTableWidget::vscrollbarChanged -> value=401967, numVisibleRows=15, modelRowCount=400000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 400000, 50000;
dataTableHeight = 429, to = 13412
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=15.
dataTableHeight = 429, to = 14104
dataTableHeight = 429, to = 14226
dataTableHeight = 429, to = 14308
dataTableHeight = 429, to = 14471
dataTableHeight = 429, to = 14552
dataTableHeight = 429, to = 14633
dataTableHeight = 429, to = 14715
dataTableHeight = 429, to = 14878
dataTableHeight = 429, to = 15000
ExtendedTableWidget::vscrollbarChanged -> value=452059, numVisibleRows=14, modelRowCount=450000.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 450000, 50000;
dataTableHeight = 429, to = 15081
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
dataTableHeight = 429, to = 15489
dataTableHeight = 429, to = 15611
dataTableHeight = 429, to = 15692
dataTableHeight = 429, to = 15774
dataTableHeight = 429, to = 15855
dataTableHeight = 429, to = 16018
dataTableHeight = 429, to = 16100
dataTableHeight = 429, to = 16140
dataTableHeight = 429, to = 16303
dataTableHeight = 429, to = 16385
dataTableHeight = 429, to = 16466
dataTableHeight = 429, to = 16548
dataTableHeight = 429, to = 16670
```

**Comment by MKleusberg** [comment]
Next try:
```diff
diff --git a/src/ExtendedTableWidget.cpp b/src/ExtendedTableWidget.cpp
index 42c9b1f..ee8b439 100644
--- a/src/ExtendedTableWidget.cpp
+++ b/src/ExtendedTableWidget.cpp
@@ -315,6 +315,8 @@ void ExtendedTableWidget::keyPressEvent(QKeyEvent* event)
         QTableView::keyPressEvent(event);
 }
 
+#include "MainWindow.h"
+
 void ExtendedTableWidget::updateGeometries()
 {
     // Call the parent implementation first - it does most of the actual logic
@@ -326,7 +328,10 @@ void ExtendedTableWidget::updateGeometries()
         // If so and if it is a SqliteTableModel and if the parent implementation of this method decided that a scrollbar is needed, update its maximum value
         SqliteTableModel* m = qobject_cast<SqliteTableModel*>(model());
         if(m && verticalScrollBar()->maximum())
+        {
+            dynamic_cast<MainWindow*>(model()->parent())->logSql(QString("ExtendedTableWidget::updateGeometries -> totalRowCount=%1, numVisibleRows=%2.").arg(m->totalRowCount()).arg(numVisibleRows()), kLogMsg_App);
             verticalScrollBar()->setMaximum(m->totalRowCount() - numVisibleRows() + 1);
+        }
     }
 }
 
@@ -337,8 +342,14 @@ void ExtendedTableWidget::vscrollbarChanged(int value)
         return;
 
     // Fetch more data from the DB if necessary
-    if((value + numVisibleRows()) >= model()->rowCount() && model()->canFetchMore(QModelIndex()))
+    int nvr = numVisibleRows();
+    if((value + nvr) >= model()->rowCount() && model()->canFetchMore(QModelIndex()))
+    {
+        dynamic_cast<MainWindow*>(model()->parent())->logSql(QString("ExtendedTableWidget::vscrollbarChanged -> value=%1, numVisibleRows=%2, modelRowCount=%3.").arg(value).arg(nvr).arg(model()->rowCount()), kLogMsg_App);
+        dynamic_cast<MainWindow*>(model()->parent())->logSql(QString("ExtendedTableWidget::vscrollbarChanged -> position_before=%1.").arg(verticalScrollBar()->value()), kLogMsg_App);
         model()->fetchMore(QModelIndex());
+        dynamic_cast<MainWindow*>(model()->parent())->logSql(QString("ExtendedTableWidget::vscrollbarChanged -> position_after=%1.").arg(verticalScrollBar()->value()), kLogMsg_App);
+    }
 }
 
 int ExtendedTableWidget::numVisibleRows()
```

**Comment by justinclift** [comment]
Hmmm, looking at my output first (I'll try your patch in a second), that visualIndexAt() call is definitely returning incorrect values.

**Comment by justinclift** [comment]
k, I'll try your patch next.  Gimme a minute. :smile:

**Comment by MKleusberg** [comment]
The dataTableHeight value in your log is constant. No idea if it's correct but it looks plausible. And it definitely should be constant as long as you don't resize the window. So as far as that goes it looks all good. The to value is wrong of course, but that's ok as long it's correct in relation to the from value: it should be the from value plus 14 or 15 in your case.

**Comment by innermous** [comment]
Why not use qobject_casts?

**Comment by justinclift** [comment]
Yeah, the "from" value is 16658, the "to" value is 16670.  No worries with that then. :wink:

@MKleusberg Output from your patch:

```
PRAGMA foreign_keys = "1";
SELECT type,name,sql,tbl_name FROM sqlite_master UNION SELECT type,name,sql,tbl_name FROM sqlite_temp_master;
PRAGMA encoding
SELECT COUNT(*) FROM (SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC);
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 0, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=53759, numVisibleRows=15, modelRowCount=50000.
ExtendedTableWidget::vscrollbarChanged -> position_before=53759.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 50000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=53759.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=15.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=15.
ExtendedTableWidget::vscrollbarChanged -> value=101408, numVisibleRows=14, modelRowCount=100000.
ExtendedTableWidget::vscrollbarChanged -> position_before=101408.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 100000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=101408.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=150280, numVisibleRows=14, modelRowCount=150000.
ExtendedTableWidget::vscrollbarChanged -> position_before=150280.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 150000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=150280.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=204038, numVisibleRows=14, modelRowCount=200000.
ExtendedTableWidget::vscrollbarChanged -> position_before=204038.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 200000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=204038.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=251688, numVisibleRows=14, modelRowCount=250000.
ExtendedTableWidget::vscrollbarChanged -> position_before=251688.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 250000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=251688.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=303003, numVisibleRows=14, modelRowCount=300000.
ExtendedTableWidget::vscrollbarChanged -> position_before=303003.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 300000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=303003.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=351874, numVisibleRows=14, modelRowCount=350000.
ExtendedTableWidget::vscrollbarChanged -> position_before=351874.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 350000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=351874.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=405633, numVisibleRows=14, modelRowCount=400000.
ExtendedTableWidget::vscrollbarChanged -> position_before=405633.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 400000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=405633.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=452060, numVisibleRows=14, modelRowCount=450000.
ExtendedTableWidget::vscrollbarChanged -> position_before=452060.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 450000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=452060.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
```

**Comment by MKleusberg** [comment]
Nope, this didn't reveal anything :crying_cat_face: I'm running out of ideas now. But let's try this slightly modified patch:
```diff
diff --git a/src/ExtendedTableWidget.cpp b/src/ExtendedTableWidget.cpp
index 42c9b1f..b823803 100644
--- a/src/ExtendedTableWidget.cpp
+++ b/src/ExtendedTableWidget.cpp
@@ -315,6 +315,8 @@ void ExtendedTableWidget::keyPressEvent(QKeyEvent* event)
         QTableView::keyPressEvent(event);
 }
 
+#include "MainWindow.h"
+
 void ExtendedTableWidget::updateGeometries()
 {
     // Call the parent implementation first - it does most of the actual logic
@@ -326,7 +328,10 @@ void ExtendedTableWidget::updateGeometries()
         // If so and if it is a SqliteTableModel and if the parent implementation of this method decided that a scrollbar is needed, update its maximum value
         SqliteTableModel* m = qobject_cast<SqliteTableModel*>(model());
         if(m && verticalScrollBar()->maximum())
+        {
+            qobject_cast<MainWindow*>(model()->parent())->logSql(QString("ExtendedTableWidget::updateGeometries -> totalRowCount=%1, numVisibleRows=%2.").arg(m->totalRowCount()).arg(numVisibleRows()), kLogMsg_App);
             verticalScrollBar()->setMaximum(m->totalRowCount() - numVisibleRows() + 1);
+        }
     }
 }
 
@@ -337,8 +342,14 @@ void ExtendedTableWidget::vscrollbarChanged(int value)
         return;
 
     // Fetch more data from the DB if necessary
-    if((value + numVisibleRows()) >= model()->rowCount() && model()->canFetchMore(QModelIndex()))
+    int nvr = numVisibleRows();
+    if((value + nvr) >= model()->rowCount() && model()->canFetchMore(QModelIndex()))
+    {
+        qobject_cast<MainWindow*>(model()->parent())->logSql(QString("ExtendedTableWidget::vscrollbarChanged -> value=%1, numVisibleRows=%2, modelRowCount=%3.").arg(value).arg(nvr).arg(model()->rowCount()), kLogMsg_App);
+        qobject_cast<MainWindow*>(model()->parent())->logSql(QString("ExtendedTableWidget::vscrollbarChanged -> position_before=%1.").arg(verticalHeader()->visualIndexAt(0)), kLogMsg_App);
         model()->fetchMore(QModelIndex());
+        qobject_cast<MainWindow*>(model()->parent())->logSql(QString("ExtendedTableWidget::vscrollbarChanged -> position_after=%1.").arg(verticalHeader()->visualIndexAt(0)), kLogMsg_App);
+    }
 }
 
 int ExtendedTableWidget::numVisibleRows()
```

In this I've also changed the case to qobject_cast for Vlad :wink:

**Comment by justinclift** [comment]
No worries, will try in a second.

Does it help that I can breakpoint on things, and potentially go poking around for values to look at? eg:

![screen shot 2016-10-08 at 18 27 01](https://cloud.githubusercontent.com/assets/406299/19214838/fb5109e6-8d84-11e6-9993-a8d935f0fbb0.png)

**Comment by justinclift** [comment]
Output from the new patch:

```
PRAGMA foreign_keys = "1";
SELECT type,name,sql,tbl_name FROM sqlite_master UNION SELECT type,name,sql,tbl_name FROM sqlite_temp_master;
PRAGMA encoding
SELECT COUNT(*) FROM (SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC);
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 0, 50000;
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=51315, numVisibleRows=14, modelRowCount=50000.
ExtendedTableWidget::vscrollbarChanged -> position_before=1710.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 50000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=1710.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=106295, numVisibleRows=14, modelRowCount=100000.
ExtendedTableWidget::vscrollbarChanged -> position_before=3543.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 100000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=3543.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=155167, numVisibleRows=14, modelRowCount=150000.
ExtendedTableWidget::vscrollbarChanged -> position_before=5172.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 150000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=5172.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=202816, numVisibleRows=14, modelRowCount=200000.
ExtendedTableWidget::vscrollbarChanged -> position_before=6760.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 200000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=6760.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=256575, numVisibleRows=14, modelRowCount=250000.
ExtendedTableWidget::vscrollbarChanged -> position_before=8552.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 250000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=8552.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=316442, numVisibleRows=14, modelRowCount=300000.
ExtendedTableWidget::vscrollbarChanged -> position_before=10548.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 300000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=10548.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=362870, numVisibleRows=14, modelRowCount=350000.
ExtendedTableWidget::vscrollbarChanged -> position_before=12095.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 350000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=12095.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
ExtendedTableWidget::vscrollbarChanged -> value=411741, numVisibleRows=15, modelRowCount=400000.
ExtendedTableWidget::vscrollbarChanged -> position_before=13724.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 400000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=13724.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=15.
ExtendedTableWidget::vscrollbarChanged -> value=456947, numVisibleRows=14, modelRowCount=450000.
ExtendedTableWidget::vscrollbarChanged -> position_before=15231.
SELECT `_rowid_`,* FROM `download_counts` ORDER BY `_rowid_` ASC LIMIT 450000, 50000;
ExtendedTableWidget::vscrollbarChanged -> position_after=15231.
ExtendedTableWidget::updateGeometries -> totalRowCount=499723, numVisibleRows=14.
```

The position_* values seem like we're getting closer. :smile:

**Comment by MKleusberg** [comment]
The only question is why they aren't even close to the ```value``` value.

Can you try setting the Prefetch value in the Preferences dialog to 500000 and see what happens? If there are still problems it's definitely not the prefetching that's causing trouble here.

**Comment by justinclift** [comment]
Same problem.  Damn. 😦

**Comment by MKleusberg** [comment]
And there are no more appearances of ```ExtendedTableWidget::vscrollbarChanged``` lines in the log I guess? Also I'm now online in Gitter, so we can chat there :smiley:

**Comment by MarqueIV** [comment]
Just jumping in here with zero experience with this code, but reading the thread so far, could it be an issue of an internal window's frame not being properly constrained by the outer 'viewport' window (and by 'window' I mean control, whatever).  I've seen cases where people have tried implementing their own scrolling mechanism and the math has gotten mixed up between what you're seeing, what you're supposed to see, and what was actually going on behind the scenes, but was obscured visually.

Again, just a complete guess, but based on something similar I'd seen in the past, so again, check the actual bounds of the various windows/frames.

**Comment by justinclift** [comment]
@MarqueIV Makes sense to me. :smile:

**Comment by MKleusberg** [comment]
I was thinking about something similar, too. But it looks like we're not really interfering with the scrolling in any way. As long as the window isn't resized and there are no prefetches to be made (we disabled that by setting the block size to 500k), we're not doing anything that would affect scrolling. At least I think so.

**Comment by innermous** [comment]
@MKleusberg ahaha :) I was just saying.

**Comment by justinclift** [comment]
🎉   Fixed by @revolter in https://github.com/sqlitebrowser/sqlitebrowser/commit/e5bca9d2dbeb92cc18cda9d0d217a8480e871a08. 🎉  😁

**Comment by MarqueIV** [comment]
So is the fact that the other bug was closed also saying this one is now too?  If so, when are the next binaries going to be made available?

**Comment by justinclift** [comment]
Well, there are the nightly builds.

Are you waiting on something in particular? :smile:

There is a possibility of us doing a 3.9.2 release in the near future, with stuff like this + memory leaks and some other minor stuff included.  But, that's not guaranteed to happen, it's just an idea being considered. :smile:

**Comment by MarqueIV** [comment]
No, no rush. Just wondering! :)

[Sent from my... wait... does anyone really read this part?!?]

**Comment by justinclift** [comment]
No worries at all. :smile:

Aside from a potential 3.9.2 release, the next major release will be 3.10.0, and we don't really have a firm timeline for that yet.  The main thing's we're wanting to get into 3.10.x are better [foreign key/constraint support](https://github.com/sqlitebrowser/sqlitebrowser/projects/2), and also "reasonably ok" integration with our new [DBHub.io](https://dbhub.io) project.

Coding on DBHub.io hasn't even begun, so it's hard to tell.  Probably have the very first alpha(s) in a few weeks though (~start of November?), mostly to get people testing DBHub integration + report back bugs / suggestions. :smile:
